sandbox/cgroup/kill.go

```diff
@@ -0,0 +1,97 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2024 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package cgroup
+
+import (
+	"fmt"
+	"os"
+	"path/filepath"
+	"syscall"
+)
+
+var KillSnapProcesses = killSnapProcessesImplV1
+
+func pickKillV1Impl() {
+	KillSnapProcesses = killSnapProcessesImplV1
+}
+
+func pickKillV2Impl() {
+	KillSnapProcesses = killSnapProcessesImplV2
+}
+
+var syscallKill = syscall.Kill
+
+func killProcessesInCgroup(dir string, signal syscall.Signal) error {
+	pids, err := pidsInFile(filepath.Join(dir, "cgroup.procs"))
+	if err != nil {
+		return err
+	}
+
+	var firstErr error
+	for _, pid := range pids {
```

Comment:
Can you leave a TODO that this is racy and that instead we should open each pid, verify the cgroup it inhabits and then kill it only if the cgroup still matches. As written we have no guarantee that the PID we are racing to kill is the one from the cgroup due to the silly design in old Linux that has recently finally been fixed with pidfd.

Reply:
I did some digging into this, this is what I found:
- For v1:
  - The pattern of `freeze -> kill -> thaw` is race free because the v1 freezer puts processes in an uninterruptable sleep, so they cannot die even if SIGKILL was sent externally.
- For v2:
  - It seems that the v2 freezer was only available since [Linux 5.2](https://kernelnewbies.org/Linux_5.2#Add_freezer_controller_for_cgroups_v2) so freezing is a no-op before 5.2
  - Also, freezing doesn't put processes in an uninterruptable sleep, so they can be killed externally and have their pid reused
  - The good news is that `pidfd_send_signal` is supported in [Linux 5.1](https://kernelnewbies.org/Linux_5.1#Safe_signal_delivery_in_presence_of_PID_reuse) so starting from 5.2 we can reliably use `freezing + pidfds` but I am not sure it is worth the extra complexity as `cgroup.kill` for v2 was introduced in [Linux 5.14](https://kernelnewbies.org/Linux_5.14#cgroup_kill_interface_to_kill_all_processes) and solves all our problems on relatively newer kernels (after Aug 2021).
  - We now have this window between [Linux 4.5](https://kernelnewbies.org/Linux_4.5#cgroup_unified_hierarchy_is_considered_stable) where v2 was introduced as stable and  [Linux 5.2](https://kernelnewbies.org/Linux_5.2#Add_freezer_controller_for_cgroups_v2) where v2 freezer was introduced where it seems it is unavoidable to have a race if v2 was used even if we add the pidfds complexity.

A note on distributions that enable v2 by default according to [runc](https://github.com/opencontainers/runc/blob/main/docs/cgroup-v2.md#cgroup-v2):
- Fedora (since 31): **31 is EOL, The currently supported 39 uses Linux 6.5, so we are good there**
- Arch Linux (since April 2021): **Rolling release, most likely using latest kernel**
- openSUSE Tumbleweed (since c. 2021): **Rolling release, currently at [6.10.4](https://software.opensuse.org/package/kernel-default), openSUSE Leap 15.5 uses [5.14](https://en.opensuse.org/Features_15.5#Linux_kernel) while openSUSE Leap 15.6 uses [6.4](https://news.opensuse.org/2024/03/07/leap-reaches-beta-phase/)**
- Debian GNU/Linux (since 11): **Debian 11 uses [Linux 5.10](https://www.debian.org/releases/bullseye/amd64/release-notes/ch-whats-new.en.html#major-packages) so we would be missing `cgroup.kill`, Debian 12 uses [6.1](https://www.debian.org/releases/bookworm/amd64/release-notes/ch-whats-new.en.html#major-packages) though**
- Ubuntu (since 21.10): **21.10 is EOL, 22.04 uses [Linux 5.15](https://ubuntu.com/kernel/lifecycle) so we would have `cgroup.kill`**
- RHEL and RHEL-like distributions (since 9): **Centos 9 uses Linux 5.14**
- Amazon Linux: **AML2 ships with both [4.15 and (default) 5.10](https://docs.aws.amazon.com/linux/al2/ug/aml2-kernel.html) (EOL 2025-06-30) while AML2023 brings [6.1](https://docs.aws.amazon.com/linux/al2023/ug/compare-with-al2-kernel.html)**

A similar discussion can be found in [systemd](https://github.com/systemd/systemd/issues/13101).

@pedronis, @zyga What are your thoughts?

Reply:
My conclusion is using `pidfds` + `pidfd_send_signal` + `confirm pidfd belongs to cgroup` would solve the race for v2 in the real world since most distros that default to using v2 use a 5.2+ kernel and anything 5.14+ will fallback to the simpler `cgroup.kill`.

I think now the question is, should we add this complexity?

Reply:
Actually we could tackle the window between Linux 4.5 and 5.2 where we cannot freeze a cgroup by doing multiple iterations when SIGKILL is passed  until the cgroup is drained of all procs even if the process keeps forking similar to what systemd is doing [here](https://github.com/systemd/systemd/blob/98b86d294adcca9887d0676190e7f697f45f7f54/src/basic/cgroup-util.c#L386-L387).

But that would also mean adding to already existing complexity of pidfds.

Reply:
I think this is a fair assesment and I would suggest that we use `cgroup.kill` and fall back to iterating over processes like we do, with a task to perhaps add pidfd support as a separate task we file in Jira and schedule sometime during this year but not for 2.65.